examples/wasm.cpp

```diff
@@ -0,0 +1,112 @@
+#include <cstdio>
+
+#include <jsapi.h>
+#include <js/CompilationAndEvaluation.h>
+#include <js/SourceText.h>
+#include <js/WasmModule.h>
+#include <js/ArrayBuffer.h>
+
+#include "boilerplate.h"
+
+// This example illustrates usage of WebAssembly JS API via embedded
+// SpiderMonkey. It does no error handling and simply exits if something
+// goes wrong.
+//
+// See 'boilerplate.cpp' for the parts of this example that are reused in many
+// simple embedding examples.
+//
+// To use the WebAssembly JIT you need to create a context and a global object,
+// and  do some setup on both of these. You also need to enter a "realm"
+// (environment within one global object) before you can execute code.
+
+/*
+hi.wat:
+(module
+  (import "env" "bar" (func $bar (param i32) (result i32)))
+  (func (export "foo") (result i32)
+    i32.const 42
+    call $bar
+  ))
+*/
+unsigned char hi_wasm[] = {
+  0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, 0x01, 0x0a, 0x02, 0x60,
+  0x01, 0x7f, 0x01, 0x7f, 0x60, 0x00, 0x01, 0x7f, 0x02, 0x0b, 0x01, 0x03,
+  0x65, 0x6e, 0x76, 0x03, 0x62, 0x61, 0x72, 0x00, 0x00, 0x03, 0x02, 0x01,
+  0x01, 0x07, 0x07, 0x01, 0x03, 0x66, 0x6f, 0x6f, 0x00, 0x01, 0x0a, 0x08,
+  0x01, 0x06, 0x00, 0x41, 0x2a, 0x10, 0x00, 0x0b
+};
+unsigned int hi_wasm_len = 56;
+
+static bool BarFunc(JSContext* cx, unsigned argc, JS::Value* vp) {
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+  args.rval().setInt32(args[0].toInt32());
+  return true;
+}
+
+static bool WasmExample(JSContext* cx) {
+  JS::RootedObject global(cx, boilerplate::CreateGlobal(cx));
+  if (!global) {
+    return false;
+  }
+
+  JSAutoRealm ar(cx, global);
+
+  // Get WebAssembly.Module and WebAssembly.Instance constructors.
+  JS::RootedValue wasm(cx);
+  JS::RootedValue wasmModule(cx);
+  JS::RootedValue wasmInstance(cx);
+  if (!JS_GetProperty(cx, global, "WebAssembly", &wasm)) return false;
+  JS::RootedObject wasmObj(cx, &wasm.toObject());
+  if (!JS_GetProperty(cx, wasmObj, "Module", &wasmModule)) return false;
+  if (!JS_GetProperty(cx, wasmObj, "Instance", &wasmInstance)) return false;
+
+
+  // Construct Wasm module from bytes.
+  JS::RootedObject module_(cx);
+  {
+    JS::RootedValueArray<1> args(cx);
+    args[0].setObject(
+      *JS::NewArrayBufferWithUserOwnedContents(cx, hi_wasm_len, hi_wasm));
```

Comment:
The return value of `JS::NewArrayBufferWithUserOwnedContents()` should be checked for null and bail out if so.